# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/wasm/WasmBuiltins.cpp
# Commit: 662cb63c42d7
# Full Hash: 662cb63c42d7874a79256a6edf1a03d83e43a014
# Author: Ryan Hunt <rhunt@eqrion.net>
# Date: 2023-08-05 09:19:01
# Regressor Bug: 1692065
# File Overlap Count: 2
# Description:
#   Bug 1692065 - wasm: Create a unique wasm::AnyRef type. r=yury,jonco
#   
#   Right now wasm references area represented as a JSObject*. This is assumed in
#   many places in SpiderMonkey, most prominently in our GC integration
#   around tracing, the store buffer, and write barriers. This will cause issues
# ==============================================================================

diff -r 8b506ab41451 -r 662cb63c42d7 js/src/wasm/WasmBuiltins.cpp
--- a/js/src/wasm/WasmBuiltins.cpp	Fri Aug 04 13:54:57 2023 +0000
+++ b/js/src/wasm/WasmBuiltins.cpp	Fri Aug 04 14:06:49 2023 +0000
@@ -639,7 +639,7 @@
 
         cx->clearPendingException();
         RootedAnyRef ref(cx, AnyRef::null());
-        if (!BoxAnyRef(cx, exn, &ref)) {
+        if (!AnyRef::fromJSValue(cx, exn, &ref)) {
           MOZ_ASSERT(cx->isThrowingOutOfMemory());
           hasCatchableException = false;
           continue;
@@ -865,7 +865,7 @@
   JSContext* cx = TlsContext.get();  // Cold code
   RootedValue val(cx, *rawVal);
   RootedAnyRef result(cx, AnyRef::null());
-  if (!BoxAnyRef(cx, val, &result)) {
+  if (!AnyRef::fromJSValue(cx, val, &result)) {
     return nullptr;
   }
   return result.get().forCompiledCode();
@@ -916,14 +916,14 @@
       case ValType::Ref: {
         // Guarded against by temporarilyUnsupportedReftypeForEntry()
         MOZ_RELEASE_ASSERT(funcType.args()[i].refType().isExtern());
-        // Leave Object and Null alone, we will unbox inline.  All we need
-        // to do is convert other values to an Object representation.
-        if (!arg.isObjectOrNull()) {
-          RootedAnyRef result(cx, AnyRef::null());
-          if (!BoxAnyRef(cx, arg, &result)) {
+        // Perform any fallible boxing that may need to happen so that the JIT
+        // code does not need to.
+        if (AnyRef::valueNeedsBoxing(arg)) {
+          JSObject* boxedValue = AnyRef::boxValue(cx, arg);
+          if (!boxedValue) {
             return false;
           }
-          argv[i].setObject(*result.get().asJSObject());
+          argv[i] = ObjectOrNullValue(boxedValue);
         }
         break;
       }